src/data/polynomial/laurent.lean

```diff
@@ -0,0 +1,314 @@
+begin
+  apply (to_finsupp_iso R).injective,
+  -- the combination `rw ← single_eq_C_mul_T, simp [-single_eq_C_mul_T, trunc], refl`
+  -- closes the goal, but involves a non-terminal `simp`.
+  rw [← single_eq_C_mul_T, trunc, add_monoid_hom.coe_comp, function.comp_app,
+    map_domain.add_monoid_hom_apply, to_finsupp_iso_apply, map_domain_single, to_finsupp_iso_apply,
+    to_finsupp_monomial],
+  refl,
+end
+
+lemma _root_.polynomial.to_laurent_injective :
+  function.injective (polynomial.to_laurent : R[X] → R[T;T⁻¹]) :=
+begin
+  refine function.has_left_inverse.injective ⟨trunc, λ f, f.induction_on' _ _⟩,
+  { exact λ f g hf hg, by simp only [hf, hg, _root_.map_add] },
+  { exact λ n r, by simp only [monomial_eq_C_mul_T, trunc_C_mul_T, int.to_nat_coe_nat] }
+end
+
+@[simp]
+lemma _root_.polynomial.to_laurent_inj (f g : R[X]) :
+  f.to_laurent = g.to_laurent ↔ f = g :=
+⟨λ h, polynomial.to_laurent_injective h, λ h, congr_arg _ h⟩
+
+lemma exists_T_pow (f : R[T;T⁻¹]) :
+  ∃ (n : ℕ) (f' : R[X]), f'.to_laurent = f * T n :=
+begin
+  apply f.induction_on' _ (λ n a, _); clear f,
+  { rintros f g ⟨m, fn, hf⟩ ⟨n, gn, hg⟩,
+    by_cases h : m ≤ n;
+    refine ⟨m + n, fn * X ^ n + gn * X ^ m, _⟩;
+    simp only [hf, hg, add_mul, add_comm (n : ℤ), int.coe_nat_add, _root_.map_add,
+      _root_.map_mul, map_pow, X_eq_T, T_pow, mul_one, mul_T_assoc]  },
+  { cases n with n n,
+    { exact ⟨0, polynomial.C a * X ^ n, by simp⟩ },
+    { refine ⟨n + 1, polynomial.C a, _⟩,
+      simp only [int.neg_succ_of_nat_eq, neg_add_self, int.coe_nat_succ, mul_T_assoc, T_zero,
+        mul_one, C_eq_C] } }
+end
+
+/--  This version of `exists_T_pow` can be called as `rcases f.exists_T_pow' with ⟨n, f', rfl⟩`. -/
+lemma exists_T_pow' (f : R[T;T⁻¹]) :
+  ∃ (n : ℕ) (f' : R[X]), f = f'.to_laurent * T (- n) :=
+begin
+  rcases f.exists_T_pow with ⟨n, f', hf⟩,
+  exact ⟨n, f', by simp [hf]⟩,
+end
+
+/--  Suppose that `Q` is a statement about Laurent polynomials such that
+* `Q` is true on *ordinary* polynomials;
+* `Q (f * T)` implies `Q f`;
+it follow that `Q` is true on all Laurent polynomials. -/
+lemma proprop (f : R[T;T⁻¹]) {Q : R[T;T⁻¹] → Prop}
+  (Qf : ∀ (f : R[X]), Q f.to_laurent)
+  (QT : ∀ f, Q (f * T 1) → Q f) :
+  Q f :=
+begin
+  rcases f.exists_T_pow' with ⟨n, f', rfl⟩,
+  induction n with n hn,
+  { simpa using Qf _ },
+  { convert QT _ _,
+    simpa using hn }
+end
+
+instance : module R[X] R[T;T⁻¹] :=
+{ smul      := λ f l, f.to_laurent * l,
+  one_smul  := λ f, by simp only [polynomial_one_eq_one, one_mul],
+  mul_smul  := λ f g l, by simp only [mul_assoc, _root_.map_mul],
+  smul_add  := λ f x y, by simp only [mul_add],
+  smul_zero := λ f, by simp only [mul_zero],
+  add_smul  := λ f g x, by simp only [add_mul, _root_.map_add],
+  zero_smul := λ f, by simp only [_root_.map_zero, zero_mul] }
+
+end semiring
+
+/-
+section comm_semiring
+variable [comm_semiring R]
+
+instance : algebra R[X] R[T;T⁻¹] :=
```

Comment:
This looks pretty scary to me; it probably creates a diamond with the case when working with Laurent polynomials over polynomials

Reply:
Ok, good thing that it is commented then!  😉

I can change it to a def or simply ignore it for now, and resurrect it/rethink it, if it becomes desirable later.